# Day 4 Reading Journal

This journal includes several required exercises, but it is meant to encourage active reading more generally.  You should use the journal to take detailed notes, catalog questions, and explore the content from Think Python deeply.

Reading: Think Python Chapter 10

**Due: Monday, January 30 at 12 noon**



## [Chapter 10](http://www.greenteapress.com/thinkpython2/html/thinkpython2011.html)

You may want to review [state diagrams in Chapter 2](http://greenteapress.com/thinkpython2/html/thinkpython2003.html). [Python Tutor](http://pythontutor.com/) is also helpful for visualizing the state of your program.



### Exercise 10.3 
Write a function called `middle` that takes a list and returns a new list that contains all but the first and last elements. So `middle([1,2,3,4])` should return `[2,3]`.

In [14]:
def middle(list):
    del list[0]
    del list[-1]
    return list

middle([3,4,5,6,7])

[4, 5, 6]

### Exercise 10.4 
Write a function called `chop` that takes a list, modifies it by removing the first and last elements, and returns `None`.

What is the difference between `middle` and `chop`? Sketch out the program state or take a look at each in Python Tutor and answer the question in the Markdown cell below.

In [11]:
def chop(list):
    del list[0]
    del list[-1]
    return None

chop([1,2,3,4,5])

After framing function, del list[0] and [-1] delete first and last elements. Then, at return None returns the value of "None"

### Exercise 10.6 
Two words are anagrams if you can rearrange the letters from one to spell the other. Write a function called `is_anagram` that takes two strings and returns `True` if they are anagrams.

In [17]:
def is_anagram(list1,list2):
    if len(list1) != len(list2):
        return False
    #!= means not equal to
    for i in range(len(list1)):
        if list1.count(list1[i]) == list2.count(list2[i]):
            return True
        else: 
            return False
        

print(is_anagram('a,p,p,l,e','a,e,p,p,l'))
print(is_anagram('conrad','radcon'))
print(is_anagram('apple','ppa'))

True
True
False


### Exercise 10.8  
The (so-called) Birthday Paradox: <br /><br />
1\. Write a function called `has_duplicates` that takes a list and returns `True` if there is any element that appears more than once. It should not modify the original list.

2\. If there are 23 students in your class, what are the chances that two of you have the same birthday? Put your answer in the Markdown cell below. You can estimate this probability by generating random samples of 23 birthdays and checking for matches. Hint: you can generate random birthdays with the randint function from the [random module](https://docs.python.org/2/library/random.html).

You can read about this problem at http://en.wikipedia.org/wiki/Birthday_paradox, and you can download Allen's solution from http://greenteapress.com/thinkpython2/code/birthday.py.

In [47]:
def has_duplicates(t): 
        #returns true if two  of elements are identical
        s = t[:]
        #t is the copy of list to avoid modification
        s.sort()
        #sort the parameter list in order of element
        for i in range(len(s)-1):
            if s[i]==s[i+1]:
            #check for adjacent elements
            #since they are assorted in alphabetical orders/numerical orders, 
            #any two same letters should be adjacent
                return True
        return False
            
a = ['a','p','p','l','e']           
has_duplicates(a)

True

In [71]:
import random


def has_duplicates(t):
  
    # s is a copy of list t to avoid modifying list t 
    s = t[:]
    s.sort()

    # sort the elements and returns true if two adjacent elements are equal
    for i in range(len(s)-1):
        if s[i] == s[i+1]:
            return True
    return False


def random_birthdays(n):
       #random_birthdays return n lists of birthdays that are completely random 
    t = []
    for i in range(n):
        bday = random.randint(1, 365)
        t.append(bday)
    return t


def count_matches(num_students, num_simulations):
    #count the number of matches for number of simulations
    count = 0
    for i in range(num_simulations):
        t = random_birthdays(num_students)
        if has_duplicates(t):
            count += 1
    return count


def counting():
    num_students = 23
    num_simulations = 1000
    count = count_matches(num_students, num_simulations)
    print(count)
    
counting()

474


After 1000 simulations with 23 numbers of students, 474 of simulations had at least one match. Therefore, the probability of having at least a pair of same birthday is 47.4%

### Challenge: Exercise 10.10 (optional)

You should read [Chapter 9.1](http://www.greenteapress.com/thinkpython2/html/thinkpython2010.html) and do Exercise 1 first.

To check whether a word is in the word list, you could use the `in` operator, but it would be relatively slow because it searches through the words in order (try it).

Because the words are in alphabetical order, we can speed things up with a bisection search (also known as binary search), which is similar to what you do when you look a word up in the dictionary. You start in the middle and check to see whether the word you are looking for comes before the word in the middle of the list. If so, then you search the first half of the list the same way. Otherwise you search the second half.

Either way, you cut the remaining search space in half. If the word list has 113,809 words, it will take about 17 steps to find the word or conclude that it’s not there.

Write a function called `bisect` that takes a sorted list and a target value and returns the index of the value in the list, if it’s there, or `None` if it’s not.

Or you could read the documentation of the `bisect` module and use that! Solution: http://greenteapress.com/thinkpython2/code/inlist.py.

## Reading Journal feedback

Have any comments on this Reading Journal? Feel free to leave them [here](https://goo.gl/forms/hZqCUAi4ir7hVN6x2) and we'll read them when you submit your journal entry. This could include suggestions to improve the exercises, topics you'd like to see covered in class next time, or other feedback.

If you have Python questions or run into problems while completing the reading, you should post them to Piazza instead so you can get a quick response before your journal is submitted.